### ISSUE TITLE
From-scratch augmentations proposal

### ISSUE THREAD

**Issue by jedrzejwalega** [Appsilon/image_flow_cytometry_fine_tune#9]
As a direct successor to [fine-tuning ConvNext the authors' way](https://github.com/orgs/Appsilon/projects/154/views/1?pane=issue&itemId=88016030&issue=Appsilon%7Cimage_flow_cytometry_fine_tune%7C6) I propose we should drop their current approach and design our own from scratch. 

A look at after-augmentations train batch preview from the authors' approach ([code here)](https://github.com/Appsilon/image_flow_cytometry_fine_tune/blob/finetune/6%20-%20Appsilon/1%20-%20Machine%20learning/Fine-tuning%20ConvNext/Supervised_learning_min_max_scaling-ConvNext.py):
![Image](https://github.com/user-attachments/assets/9782d572-eb27-4dab-9434-1f1e90024c71)

- The [padding](https://github.com/marrlab/scifAI/blob/c73f6e821de818d377328c21f05dac53a7d9cf77/scifAI/dl/dataset.py) performed by scifAI's library (padding by per-image per-channel mean) varies greatly across images and channels.
- Our images are very small. After padding to the nn's input size most of the image becomes just the padding. Perhaps we can do better.
- We see weird, rectangular boxes around most cells (the original image's borders). Ideally I would like to do away with those.

Additional factors to consider:
- The cells may vary slightly in size (biologically).
- Authors use only horizontal and vertical flipping, no rotation during training (they oddly have it when calculating statistics...)
- The microscope image quality is not great in the best scenario to very blurry in the worst one.

My proposal for our own take:

1. Use an algorithm to find the cell borders, considering all channels together. Once we find them, we crop the images to this size. Never used it, but maybe [OpenCV](https://opencv.org/) can help us?
2. Apply mirror padding to the desired nn's input size. I'd like to maximize the cell information coverage in the picture, but simple resizing might distort the cells.
3. A chain of random transformations (50% each one will be applied):

- Horizontal flip
- Vertical flip
- Rotation (180 degrees)
- Random resized crop (small one, 80% of the picture always preserved) - to simulate biologically varying cell sizes

4. Simulate bad image quality (applied independently):

- Random adjust sharpness (worsen by 30% at 70% probability) - frequent simulation of mild focus issue
- Random Gaussian blur (kernel=3, std in the range=[0.15 - 0.25], probability of applying at 40%) - meant to simulate an occasional, stronger blur in the image.

5. Min max normalization

**Comment by jedrzejwalega** [comment]
![Image](https://github.com/user-attachments/assets/b38ae9f7-18c2-4905-b533-8114dfc2118c)

Above is a batch preview with no augmentations. As mentioned above, I originally thought that we'll have to find cell borders on our own since channels like brightfield (first one from the top) appeared to have no box around them. Because of this, I thought it was in a different shape than the rest of the channels. But as seen in the preview above it actually has a box, identical in size to the other channels - it was just masked by Gaussian in the original batch preview.

This simplifies things - we **don't need to find the cell borders** at all, they are already cut out for us. We can use the original images for our own augmentations.

Also, after internal discussion with @Griphu and @swiezew we have decided to **not go with mirror padding**. The reason for this is that padded, mirrored parts of an image would create duplicates of the original cell. Since the classification task involves a class called Multiplets, directly tied to the number of cells, mirror padding would likely confuse the model.

**Comment by jedrzejwalega** [comment]
The agreed approach (for a first ConvNext run) is the following:

1. Pad original images with zeroes (for uniform background across channels, unlike the mean the authors were using).
2.  50% for random vertical flip
3.  50% for random horizontal flip
4. 50% for rotation in 90 degree angle. Along with both of the flip this gives us a chance of any cell position fed to the model. The authors didn't use rotation and all, so it gives us many more possible inputs to the model compared to their approach.
5. 50% for random resized crop, zoom up to 20%. It is meant to simulate differences in the cell sizes and be non-aggressive.
6. Normalization with training set's calculated statistics for each channel.

We also have two ideas we're not implementing from the get-go, but might in the next runs:

1. Use Gaussian noise or random adjust sharpness to simulate varying image quality. It's not that high to begin with, so we're not starting off with this.
2. Use SmallestMaxSize + CenterCrop to first resize the image while keeping the proportions of the picture and then take center crop to adjust the dimensions to the net's input. This is the approach used in the pretraining of the ConvNext we will be leveraging. I'm thinking that small cells on a field full of black may confuse the model, but we'll see.

**Comment by jedrzejwalega** [comment]
Batch preview with our augments (train):

![Image](https://github.com/user-attachments/assets/8a9f749b-0a06-410c-9e69-c3f4d2c106e6)

Batch preview with our augments (val/test), which includes only normalization:

![Image](https://github.com/user-attachments/assets/28a39680-08c7-42c5-9de1-814d95e73514)

**Comment by jedrzejwalega** [comment]
Proposal set, closing for separate task for implementation.
